apollo-compiler/src/main/kotlin/com/apollographql/apollo3/compiler/TypeName.kt

```diff
@@ -0,0 +1,89 @@
+package com.apollographql.apollo3.compiler
+
+import com.apollographql.apollo3.compiler.codegen.java.JavaAnnotations
+import com.apollographql.apollo3.compiler.codegen.java.JavaClassNames
+import com.squareup.javapoet.ClassName
+import com.squareup.javapoet.CodeBlock
+import com.squareup.javapoet.ParameterizedTypeName
+import com.squareup.javapoet.TypeName
+import com.squareup.javapoet.WildcardTypeName
+
+fun TypeName.isList() =
+  (this is ParameterizedTypeName && rawType == JavaClassNames.List)
+
+fun TypeName.listParamType(): TypeName {
+  return (this as ParameterizedTypeName)
+    .typeArguments
+    .first()
+    .let { if (it is WildcardTypeName) it.upperBounds.first() else it }
+}
+
+fun TypeName.isOptional(expectedOptionalType: ClassName? = null): Boolean {
+  val rawType = (this as? ParameterizedTypeName)?.rawType ?: this
+  return if (expectedOptionalType == null) {
+    rawType == JavaClassNames.Optional || rawType == JavaClassNames.Input
```

Comment:
Compared to the original code, only `com.apollographql.apollo3.api.Optional` and `com.apollographql.apollo3.api.Input` are supported. (Input is deprecated so I'm not sure whether it's actually needed here.)

Reply:
Maybe `Input` is not needed?

Reply:
I think you're right. The generated code won't use `Input` anywhere.